src/controllers/hid/hidiothread.cpp

```diff
@@ -204,11 +205,30 @@ void HidIoThread::updateCachedOutputReportData(quint8 reportID,
 
     mapLock.unlock();
 
+    // If useNonSkippingFIFO is false, the report data are cached here
+    // If useNonSkippingFIFO is true, this cache is cleared
     actualOutputReportIterator->second->updateCachedData(
```

Comment:
Why is the mutex locked and this operation invoked even when using the FIFO strategy? The cache and the mutex should be bypassed. Ideally the dispatch strategy for each reportID would be immutable and configured at construction time. At least the FIFO report IDs need to be known in advance to avoid the locking.

Reply:
The call of updateCachedData is always needed to guarantee seamless transitions between both modes.
If a skipable report is cached, it must be cleared, because otherwise it would be send after the report from the non-skipable report from the FIFO - and would overwrite with outdated data. This happens in updateCachedData as well as the check for errors.

Reply:
I acknowledge that the current design is not suitable to for this optimization. But in principle the dispatch strategy should be fixed per report ID and not change with every call.

Reply:
Caching the mode only for the sender would avoid locking the mutex with every call. The cache only needs to be cleared once after changing the mode.

Reply:
The only datastructure per output report is currently HidIoOutputReport. Adding a additional data structure per output report would prevent the mutex locking in FIFO mode. But this mutex is only locked for the very short time of looking up the map for the map iterator of the report.
This takes nanoseconds, while the sending of a report takes several milliseconds. And in FIFO mode, calling this function must not be done in higher rate than the HID subsystem can send reports to the device. This ratio makes a mutex blocking very unlikely - and even when it happens, it will block the other thread only for several nanoseconds.
I don't think, that it makes sense to add additional code here, because the impact will be insignificant in practice.

Reply:
However short and tiny the critical section might be, you have no control over the OS scheduler. Generous use of mutexes is always a red herring.